File: tools/kwprobe/probe.rb
# The Ruby column of tools/kwprobe/run.py. One cell per row on stdout.
#
# Run with -W0: Onigmo warns on stderr about some spellings it accepts (a
# brace-less \pL, for one), and the cell is the answer, not the warning.

def decode(s)
  s.gsub(/\\u\{([0-9A-Fa-f]+)\}/) { $1.to_i(16).chr(Encoding::UTF_8) }
   .gsub('\\n', "\n").gsub('\\r', "\r").gsub('\\t', "\t").gsub('\\\\', '\\')
end

def encode(s)
  return '""' if s.empty?
  s.each_char.map do |c|
    o = c.ord
    case
    when c == '\\' then '\\\\'
    when c == "\n" then '\\n'
    when c == "\r" then '\\r'
    when c == "\t" then '\\t'
    when o < 0x20 || o > 0x7e then format('\\u{%X}', o)
    else c
    end
  end.join
end

File.foreach(ARGV[0], encoding: 'UTF-8') do |raw|
  line = raw.strip
  next if line.empty? || line.start_with?('#')
  f = line.split(' :: ').map(&:strip)
  pattern = f[1]
  subject = f.length > 2 ? decode(f[2]) : nil
  replace = nil
  split = false
  skip = false
  f[3..].to_a.each do |o|
    if o.start_with?('replace=') then replace = decode(o[8..])
    elsif o == 'split' then split = true
    elsif o.start_with?('skip=') && o[5..].split(',').include?('ruby') then skip = true
    end
  end
  if skip then puts 'n/a'; next end
  re = begin
    Regexp.new(pattern)
  rescue StandardError
    nil
  end
  if re.nil? then puts '-'; next end
  if subject.nil? then puts 'ok'; next end
  begin
    if replace
      if re.match(subject).nil? then puts 'no'
      else puts encode(subject.sub(re, replace))
      end
    elsif split
      puts '[' + subject.split(re, -1).map { |p| p.empty? ? '' : encode(p) }.join(',') + ']'
    else
      m = re.match(subject)
      puts m.nil? ? 'no' : encode(m[0])
    end
  rescue StandardError
    puts 'err'
  end
end

File: tools/kwprobe/probe.py
"""The Python column of tools/kwprobe/run.py. One cell per row on stdout."""

import re
import sys


def decode(s: str) -> str:
    out, i = [], 0
    while i < len(s):
        c = s[i]
        if c == "\\" and i + 1 < len(s):
            n = s[i + 1]
            if n == "n":
                out.append("\n"); i += 2; continue
            if n == "r":
                out.append("\r"); i += 2; continue
            if n == "t":
                out.append("\t"); i += 2; continue
            if n == "\\":
                out.append("\\"); i += 2; continue
            if n == "u" and i + 2 < len(s) and s[i + 2] == "{":
                j = s.index("}", i)
                out.append(chr(int(s[i + 3:j], 16))); i = j + 1; continue
        out.append(c); i += 1
    return "".join(out)


def encode(s: str) -> str:
    if s == "":
        return '""'
    out = []
    for c in s:
        if c == "\\":
            out.append("\\\\")
        elif c == "\n":
            out.append("\\n")
        elif c == "\r":
            out.append("\\r")
        elif c == "\t":
            out.append("\\t")
        elif ord(c) < 0x20 or ord(c) > 0x7E:
            out.append("\\u{%X}" % ord(c))
        else:
            out.append(c)
    return "".join(out)


def cell(fields: list[str]) -> str:
    pattern = fields[1]
    subject = decode(fields[2]) if len(fields) > 2 else None
    opts = fields[3:]
    replace = split = None
    for o in opts:
        if o.startswith("replace="):
            replace = decode(o[8:])
        elif o == "split":
            split = True
        elif o.startswith("skip=") and "python" in o[5:].split(","):
            return "n/a"
    try:
        rx = re.compile(pattern)
    except Exception:
        return "-"
    if subject is None:
        return "ok"
    try:
        if replace is not None:
            if rx.search(subject) is None:
                return "no"
            return encode(rx.sub(replace, subject, count=1))
        if split:
            return "[" + ",".join(encode(p) if p else "" for p in rx.split(subject)) + "]"
        m = rx.search(subject)
        return "no" if m is None else encode(m.group(0))
    except Exception:
        return "err"


def main() -> None:
    for line in open(sys.argv[1], encoding="utf-8"):
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        fields = [f.strip() for f in line.split(" :: ")]
        print(cell(fields))


main()
